proxy/coco/src/peer/subroutines.rs

```diff
@@ -0,0 +1,278 @@
+//! Management of subroutines.
+
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use futures::stream::{BoxStream, FuturesUnordered, SelectAll, StreamExt as _};
+use tokio::{
+    sync::{broadcast, mpsc},
+    time::interval,
+};
+
+use librad::{
+    net::{peer::Gossip, protocol::ProtocolEvent},
+    peer::PeerId,
+    uri::{RadUrl, RadUrn},
+};
+
+use crate::{
+    convert::MaybeFrom as _,
+    spawn_abortable::{self, SpawnAbortable},
+    state::State,
+};
+
+use super::{
+    announcement, control, gossip,
+    run_state::{
+        AnnounceInput, Command, Config as RunConfig, ControlCommand, Event, Input, RequestCommand,
+        RequestInput, RunState, SyncInput, TimeoutInput,
+    },
+    sync, RECEIVER_CAPACITY,
+};
+
+/// Management of "subroutine" tasks.
+#[allow(clippy::missing_docs_in_private_items)]
+pub struct Subroutines {
```

Comment:
If I understand this structure correctly, it does two things: It handles input events and the business logic around them and it manages tasks and concurrency. Maybe we could separate these two things more clearly. More concretely we could have one logical thread that takes input events and builds futures that are then pushed on a worker to be executed as tasks. The worker would be responsible for spawning tasks and terminating when one of the tasks errors. It would also ensure that all tasks are aborted on an error

Reply:
> It handles input events and the business logic around them and it manages tasks and concurrency

The business logic is encapsulated in the `RunState`.

> The worker would be responsible for spawning tasks and terminating when one of the tasks errors. It would also ensure that all tasks are aborted on an error

What you describe is the `Subroutines` responsibility already, I don't see what an extra worker abstraction would add to this.

Reply:
If by "business logic" you mean the morphism `Command -> SpawnAbortable`, then yeah, I also thought about how this could be abstracted out. It's not so obvious how to capture the parameters though (namely `state` and `store`), sans partial application. (Did someone say `TaskFactoryImpl`?)

Reply:
Could we expand for what we are optimising for? Do we want to be able to test this code block in isolation? Are we keen to find static guarantees for the `Command -> SpawnAbortable` mapping?

There is already quite a lot of infrastructure involved to run the peer machinery which tips the noise-to-signal ratio unfavourably, so any more of it should give us sick gains.

Reply:
Extensibility. If you want to add a new command, you need to change code in at least three places, one of which is `poll` -- which should not at all change, but the reviewer will have to look at the diff to convince herself of this being the case.

Reply:
I guess moving the pattern match to an inherent method would be good enough for that

Reply:
a087b84

Reply:
> If by "business logic" you mean the morphism `Command -> SpawnAbortable`, then yeah, I also thought about how this could be abstracted out.

Yes, this is (almost) what I meant. I was talking about `Command -> impl Future`.

> Could we expand for what we are optimising for?

Separation of concerns. The concurrency/worker logic is not easy to understand follow, so keeping it separate aids that. Moreover we could write tests for the worker more easily. Don’t let this be a blocker, though.